### PR TITLE
Update/version python

### PR DESCRIPTION
**Pull request by VascoSch92**
Hey, 

I update python, numpy and simply and it seems to work. The test are passing all but not the FGA_factorization_naive. There is a problem with the logic. I will give a look.

For instance, you can give a look to the pr ;-)

I formatted the tests but I changed nothing

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # List of requirements
-# Run 
-sympy>=1.1.1
-numpy>=1.13.1
+# Run
+python == 3.9.18
```
Comment (tommyod) [inline requirements.txt]: You cannot add python version to requirements file. The file is for packages.

File: tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def random_zero_heavy(low, high):
+    """
+    Draw a random number, with approx 50% probability of zero.
+    """
+    return choice(list(range(low, high)) + [0] * (high - low))
+
+
+def random_from_list(number, list_to_take_from):
```
Comment (tommyod) [inline tests/utils.py:12]: Isnt this more or less the same as

https://docs.python.org/3/library/random.html#random.choices

?

**Review by tommyod — COMMENTED** [review]
Mostly looks good. I left some comments. Some questions:

- what did you use to format the code?
- could you set up a github action to run the tests. we should know that tests pass before merging. i tried for a few minutes, but got an error (see the other PR). could you make an attempt? just replicate however you got the code to run locally.

File: requirements.txt (hunk 1 above)
Comment (VascoSch92) [reply requirements.txt]: Yes my bad ;-)

File: tests/utils.py (hunk 2 above)
Comment (VascoSch92) [reply tests/utils.py:12]: I didn't look in details into these methods. I just put the method in the script utils.py

But I think you could write something like (if you want to use `random.choices`):

```python
def generate_random_list(length, low, high):
    choices = [0, random.randint(low, high+1)] 
    random_list = random.choices(choices, k=length)
    
    return random_list
```

**Comment by VascoSch92** [comment]
> Mostly looks good. I left some comments. Some questions:
> 
> * what did you use to format the code?
> * could you set up a github action to run the tests. we should know that tests pass before merging. i tried for a few minutes, but got an error (see the other PR). could you make an attempt? just replicate however you got the code to run locally.

I'm just following the PEP directive about formatting. In the company where I work we use `yapf`.

Why? I hope I didn't make any gross mistakes

I can try to setup GitHub Actions ,-) but I'm not an expert...

File: tests/utils.py (hunk 2 above)
Comment (tommyod) [reply tests/utils.py:12]: My bad, I thought it did exactly the same as random.choices. If not, you can keep your implmentation, or use the one above - either way is good :)

**Comment by tommyod** [comment]
> Why? I hope I didn't make any gross mistakes

No, it looks good! I was thinking we could add the formatter to a GH action as well, to enforce the formatting rules in the future.

> I can try to setup GitHub Actions ,-) but I'm not an expert...

Please give it a try! Thanks!

**Comment by tommyod** [comment]
Now that we have an action that runs the tests, you can rebase on top of that, fix conflicts, and make changes. If tests pass I'm probably happy with changes, as long as you don't do anything too crazy :)

**Comment by VascoSch92** [comment]
> Now that we have an action that runs the tests, you can rebase on top of that, fix conflicts, and make changes. If tests pass I'm probably happy with changes, as long as you don't do anything too crazy :)

haha ok i will do ;-)

**Pull request closed without merging**